### ISSUE TITLE
Create interface for User

### ISSUE THREAD

**Issue by foxycode** [nette/security#20]
- bug report? no
- feature request? yes
- version: 3.0

Few times I wanted to use own `User` implementation, bt there is no simple way how to force Nette to use it. I'd like to have `IUser` same way as `IUserStorage` to be able to use my own implementation in framework.

**Comment by mrtnzlml** [comment]
User is [registered in DIC](https://github.com/nette/security/blob/master/src/Bridges/SecurityDI/SecurityExtension.php#L46) as a service with the name therefore it should be possible to replace it with your own implementation [similarly like in this thread](https://forum.nette.org/cs/19816-jak-jednoduse-nahradit-nette-security-user-vlastni-tridou).

**Comment by foxycode** [comment]
Problem is, there is User class used directly in Nette code. If not in code, definitelly in comments which isn't good for IDE.

**Comment by JanTvrdik** [comment]
My tip: don't use the `User` class at all, operate directly on `IUserStorage`. It has almost the same API.

**Comment by foxycode** [comment]
@JanTvrdik Thanks for the tip, but this doesn't solve issue in framework. User is somehow hardcoded and it would be great if it wasn't. I'm happy to make PR (or at least try it), but I want to know others opinions first.

**Comment by enumag** [comment]
@foxycode Actually the User class is problematic in quite a few use-cases so in advanced systems it's better to not use it at all. Better than a PR it's better to just make your own library like I did.

**Comment by foxycode** [comment]
@enumag Again, why not solve it directly in framework? Yes, I can make my own library. I actually already did.

I believe, if we put our heads together, we can solve it directly in framework. I train new programmers in Nette and it's confusing for them. I can't simply tell them to write own library :)

**Comment by enumag** [comment]
@foxycode It would become a new package separate from nette/security anyway (because of BC). In which case it doesn't matter if it is in nette namespace or not. Can't speak for @dg but I don't think he wants to add something completely new into nette.

**Comment by foxycode** [comment]
@enumag I think we can afford BC breaks in 3.0 and this is only about adding interface. Possibly this wouldn't need BC at all. I'd like to know what @dg thinks too. Can you tell me which problems do you see with User?

**Comment by JanTvrdik** [comment]
> this is only about adding interface

Yes, but it is an interface of a class with bad name and bad design. The goal should be to remove (or somehow fix?) the class instead of making even more stuff dependent on in.

**Comment by foxycode** [comment]
@JanTvrdik Ok, I agree with it :) I'd like to try, but need to know what others don't like about User class. Can you tell your opinion?

**Comment by JanTvrdik** [comment]
Few random things that are bad about the User class.

1) The name – User is usually an entity in application. The responsibility of this class is not to represent a user but to serve as some kind of manager of *active user session*.
2) It has way to many responsibilities. It is better to use directly IAuthenticator, IUserStorage and some better alternative to IAuthorizator.
3) It relies on bad interface IAuthorizator (see this [RFC on forum](https://forum.nette.org/cs/13458-security-iauthorizator-a-identita#p99180))

**Comment by dakujem** [comment]
God, I just wanted to post the same.

I am upgrading a larger codebase from nette v 2.4 and it's just not possible, since we are using an extended version of `User` class, bud suddenly some methods are `final` 😖. What is even worse, `Nette\Application\UI\Presenter` class won't accept any other than `Nette\Security\User` object and the `getUser` method is (yet again) `final`.

So... how do we extend the functionality of `User` object?? Please, this is a _framework_. Why do we constantly need to hit `private` and `final` stuff? We want to extend and bend a framework. This is where Nette gets in way. I do not wish to copypaste 1400 lines of `Presenter` code just to be able to use my own `User` implementation.

In particular, in my case I have some custom logic in `User::getIdentity` method, why it is final, I have no idea.

**Comment by sommcz** [comment]
Hi, any updates on this topic? It's quite old, but still it is a issue when upgrading large codebase from v2

**Comment by dakujem** [comment]
In case the interface is too much to ask or simply not desired, I suggest also removing the return type hint, so that the overriding methods can return anything, not descendants of the `User` object only.
```php
// Presenter.php

    /**
     * @return Nette\Security\User
     */
    public function getUser() //: Nette\Security\User
    {
        if (!$this->user) {
            throw new Nette\InvalidStateException('Service User has not been set.');
        }

        return $this->user;
    }
```
This is a backwards-compatible <10 minute fix.

**Comment by jkuchar** [comment]
Simple, stop using it at all. Only hard which is to how to remove it from DI, so nobody can wire it. I have approached it using this fake factory overriding original registration from [nette/security di-extension](https://github.com/nette/security/blob/d3588dca73d11e8517fddcd0491b5148490d8bb2/src/Bridges/SecurityDI/SecurityExtension.php#L83-L84).

```neon
services:
	security.user:
		factory: CampApp\App\Infrastructure\NoNetteUserFactory::throwErrorWhenAccessed()
		autowired: false
```

```php
final class NoNetteUserFactory
{
	public static function throwErrorWhenAccessed(): User
	{
		throw new UsageException('Hey! Nette\User is disabled in this project (as interface does not match use-cases needed in this project). Use <YourClass> instead.');
	}
}
```

**Comment by dakujem** [comment]
Wouldn't it then be better to completely remove it from `Presenter`, extension etc.? It gets in the way.
Why not just enable overriding, which is simpler and backwards-compatible?
